src/moin/templates/ajaxsearch.html

```diff
@@ -56,14 +56,25 @@ <h4>{{ _("Please check these tickets if they cover your issue:") }}</h4>
                     <span class="moin-suggestion-terms">{{ content_suggestions }}</span>
                 </p>
             {% endif %}
-
-            {% set count = [0] %}
             {% if results %}
                 <div class="moin-search-results">
-                    <table>
+                    {% set count = [0] %}
+                    <table
+                        {% set vars = {'old_acl_in': [None, None, None], 'old_permission': False} %}
                         {% for result in results %}
                             {% set fullname = result.namespace + '/' + result.name[0] if result.namespace else result.name[0] %}
-                            {% if user.may.read(fullname) %}
+                            {% set new_acl = result['acl'] if 'acl' in result else '' %}
+                            {% set new_parentnames = theme_supp.parentnames(result.name) %}
+                            {% set new_acl_in = [new_acl, result['namespace'], new_parentnames] %}
+                            {% if vars['old_acl_in'] == new_acl_in %}
+                                {# save a few seconds by avoiding recalculation of acls #}
+                                {% set permission = vars['old_permission'] %}
+                            {% else %}
+                                {% set permission = flaskg.storage.may_read_rev(result) %}
+                            {% endif %}
```

Comment:
this looks a bit like a 1-entry LRU cache?

maybe rather use a 100 entry real LRU cache? iirc there are ready-to-use implementations of this in either moin2 or stdlib.

Reply:
Opened #1079. The ACL code seems difficult to grok.